transport/handler_server.go

```diff
@@ -319,7 +320,16 @@ func (ht *serverHandlerTransport) HandleStreams(startStream func(*Stream)) {
 				s.buf.put(&recvMsg{data: buf[:n]})
 			}
 			if err != nil {
-				s.buf.put(&recvMsg{err: err})
+				if err, ok := err.(http2.StreamError); ok {
+					code := http2ErrConvTab[err.Code]
+					if code == codes.OK {
+						s.buf.put(&recvMsg{err: io.EOF})
+					} else {
+						s.buf.put(&recvMsg{err: StreamErrorf(code, "%v", err)})
+					}
+				} else {
+					s.buf.put(&recvMsg{err: io.EOF})
```

Comment:
Sure, I'll take a look. By TestPingPong_4 do you mean TestPingPong with -cpu=4? And are you able to reproduce easily on your own machine?

I can't reproduce so far, with either:

```
$ go test -v   -run=PingPong -cpu=4 ./test -only_env=handler-tls  -count=200 -race
```
or
```
$ go test -v   -run=PingPong -cpu=4 ./test -only_env=handler-tls  -count=200 
```